cpp/ql/src/semmle/code/cpp/dataflow/TaintTracking.qll

```diff
@@ -226,4 +231,35 @@ module TaintTracking {
     e instanceof AlignofOperator
   }
 
-}
\ No newline at end of file
+  private predicate exprToDefinitionByReferenceStep(Expr exprIn, Expr argOut) {
+    exists(DataFlowFunction f, Call call, FunctionOutput outModel, int argOutIndex |
+      call.getTarget() = f and
+      argOut = call.getArgument(argOutIndex) and
+      outModel.isOutParameterPointer(argOutIndex) and
+      exists(int argInIndex, FunctionInput inModel |
+        f.hasDataFlow(inModel, outModel)
```

Comment:
Why does this use of `DataFlowFunction.hasDataFlow` belong in taint tracking rather than standard data flow? Please add a test. Please also add a comment on why this is different from https://github.com/Semmle/ql/blob/45a35a8572e6ee5133e9ebdc1e3c1dae33c1153f/cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll#L301-L302. It looks to me like the difference is that this code uses `exprIn = call.getArgument(argInIndex)` instead of `call.passesByReference(argInIndex, exprIn)`.

Reply:
This handles `memcpy(&dest_var, tainted_ptr, len)`, effectively taint flowing from a pointer to a dereference.